File: apps/android/app/src/main/java/tech/stygian/presently/oauth/OAuthDiscovery.kt
package tech.stygian.presently.oauth

import org.json.JSONObject
import java.net.URI
import java.net.URLEncoder

class OAuthDiscovery(private val http: OAuthHttpClient) {
    suspend fun discover(rawIdentifier: String): DiscoveryResult {
        val identifier = rawIdentifier.trim().removePrefix("@")
        if (identifier.isEmpty()) throw OAuthException("Enter a valid handle or DID.")

        val handle: String?
        val did: String
        if (identifier.startsWith("did:")) {
            handle = null
            did = identifier
        } else {
            if (!HandlePattern.matches(identifier)) {
                throw OAuthException("Enter a valid AT Protocol handle or DID.")
            }
            handle = identifier.lowercase()
            did = resolveHandle(handle)
        }

        val didDocument = resolveDid(did)
        if (didDocument.getString("id") != did) {
            throw OAuthException("The account's DID document could not be verified.")
        }
        if (handle != null) {
            val aliases = didDocument.optJSONArray("alsoKnownAs")
            val claimsHandle = aliases != null && (0 until aliases.length()).any {
                aliases.getString(it).equals("at://$handle", ignoreCase = true)
            }
            if (!claimsHandle) {
                throw OAuthException("The account's DID document does not claim that handle.")
            }
        }
        val pdsUrl = pdsOrigin(did, didDocument)

        val resource = http.getJson("$pdsUrl/.well-known/oauth-protected-resource")
        val servers = resource.optJSONArray("authorization_servers")
        if (servers == null || servers.length() != 1) {
            throw OAuthException("The PDS returned invalid OAuth resource metadata.")
        }
        val issuer = secureOrigin(servers.getString(0))
        val metadata = http.getJson("$issuer/.well-known/oauth-authorization-server")
        val server = validateServer(metadata, issuer)

        return DiscoveryResult(identifier, did, handle, pdsUrl, server)
    }

    suspend fun validateAccount(did: String, pdsUrl: String, issuer: String) {
        val document = resolveDid(did)
        if (document.getString("id") != did || pdsOrigin(did, document) != pdsUrl) {
            throw OAuthException("The authorized account does not match its PDS.")
        }
        val resource = http.getJson("$pdsUrl/.well-known/oauth-protected-resource")
        val servers = resource.optJSONArray("authorization_servers")
        if (servers == null || servers.length() != 1 ||
            secureOrigin(servers.getString(0)) != issuer
        ) {
            throw OAuthException("The authorized account does not match its OAuth server.")
        }
    }

    private suspend fun resolveHandle(handle: String): String {
        runCatching {
            val response = http.get("https://$handle/.well-known/atproto-did")
            if (response.statusCode == 200) {
                val did = String(response.body, Charsets.UTF_8).trim()
                if (did.startsWith("did:")) return did
            }
        }

        val name = URLEncoder.encode("_atproto.$handle", Charsets.UTF_8.name())
        val response = http.get(
            "https://cloudflare-dns.com/dns-query?name=$name&type=TXT",
            mapOf("Accept" to "application/dns-json"),
        )
        if (response.statusCode == 200) {
            val answers = JSONObject(String(response.body, Charsets.UTF_8))
                .optJSONArray("Answer")
            if (answers != null) {
                for (index in 0 until answers.length()) {
                    val value = answers.getJSONObject(index)
                        .optString("data")
                        .trim('"')
                    if (value.startsWith("did=")) {
                        return value.removePrefix("did=")
                    }
                }
            }
        }
        throw OAuthException("That handle could not be resolved.")
    }

    private suspend fun resolveDid(did: String): JSONObject {
        val url = when {
            did.startsWith("did:plc:") -> "https://plc.directory/$did"
            did.startsWith("did:web:") -> {
                val parts = did.removePrefix("did:web:").split(":")
                val host = java.net.URLDecoder.decode(
                    parts.first(),
                    Charsets.UTF_8.name(),
                )
                if (host.contains("/") || host.contains("@") || host.contains("\\")) {
                    throw OAuthException("The account's did:web identifier is invalid.")
                }
                if (parts.size == 1) {
                    "https://$host/.well-known/did.json"
                } else {
                    val path = parts.drop(1).joinToString("/") {
                        java.net.URLDecoder.decode(it, Charsets.UTF_8.name())
                    }
                    "https://$host/$path/did.json"
                }
            }
            else -> throw OAuthException("Presently supports did:plc and did:web accounts.")
        }
        return http.getJson(url)
    }

    private fun pdsOrigin(did: String, document: JSONObject): String {
        val services = document.optJSONArray("service")
            ?: throw OAuthException("The account does not declare an AT Protocol PDS.")
        for (index in 0 until services.length()) {
            val service = services.getJSONObject(index)
            val id = service.optString("id")
            if ((id == "#atproto_pds" || id == "$did#atproto_pds") &&
                service.optString("type") == "AtprotoPersonalDataServer"
            ) {
                return secureOrigin(service.getString("serviceEndpoint"))
            }
        }
        throw OAuthException("The account does not declare an AT Protocol PDS.")
    }

    private fun validateServer(value: JSONObject, expectedIssuer: String): AuthorizationServer {
        fun strings(name: String) = value.optJSONArray(name)?.strings().orEmpty()
        val issuer = value.optString("issuer")
        val authorizationEndpoint = value.optString("authorization_endpoint")
        val tokenEndpoint = value.optString("token_endpoint")
        val parEndpoint = value.optString("pushed_authorization_request_endpoint")
        val valid = issuer == expectedIssuer &&
            strings("response_types_supported").contains("code") &&
            strings("grant_types_supported").containsAll(
                listOf("authorization_code", "refresh_token"),
            ) &&
            strings("code_challenge_methods_supported").contains("S256") &&
            strings("token_endpoint_auth_methods_supported").contains("none") &&
            strings("scopes_supported").contains("atproto") &&
            value.optBoolean("authorization_response_iss_parameter_supported") &&
            value.optBoolean("require_pushed_authorization_requests") &&
            strings("dpop_signing_alg_values_supported").contains("ES256") &&
            value.optBoolean("client_id_metadata_document_supported") &&
            isSecureEndpoint(authorizationEndpoint) &&
            isSecureEndpoint(tokenEndpoint) &&
            isSecureEndpoint(parEndpoint)
        if (!valid) {
            throw OAuthException(
                "The authorization server does not meet AT Protocol OAuth requirements.",
            )
        }
        return AuthorizationServer(
            issuer,
            authorizationEndpoint,
            tokenEndpoint,
            parEndpoint,
        )
    }

    private fun secureOrigin(value: String): String {
        val uri = URI(value)
        if (uri.scheme != "https" || uri.host == null || uri.userInfo != null) {
            throw OAuthException("OAuth servers must use a secure HTTPS origin.")
        }
        return URI("https", null, uri.host, uri.port, null, null, null).toString()
    }

    private fun isSecureEndpoint(value: String): Boolean = runCatching {
        val uri = URI(value)
        uri.scheme == "https" && uri.host != null && uri.userInfo == null
    }.getOrDefault(false)

    companion object {
        private val HandlePattern = Regex(
            """(?i)^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z](?:[a-z0-9-]{0,61}[a-z0-9])?$""",
        )
    }
}

File: apps/ios/Presently/OAuth/OAuthDiscovery.swift
import Foundation

struct OAuthDiscovery {
    struct Result: Sendable {
        let inputIdentifier: String
        let did: String
        let handle: String?
        let pdsURL: URL
        let server: AuthorizationServerMetadata
    }

    private let http: OAuthHTTPClient

    init(http: OAuthHTTPClient) {
        self.http = http
    }

    func discover(identifier rawIdentifier: String) async throws -> Result {
        var identifier = rawIdentifier
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if identifier.hasPrefix("@") {
            identifier.removeFirst()
        }

        guard !identifier.isEmpty else {
            throw OAuthValidationError.invalidIdentifier
        }

        let handle: String?
        let did: String
        if identifier.hasPrefix("did:") {
            handle = nil
            did = identifier
        } else {
            guard isValidHandle(identifier) else {
                throw OAuthValidationError.invalidIdentifier
            }
            handle = identifier.lowercased()
            did = try await resolveHandle(handle!)
        }

        let didDocument = try await resolveDID(did)
        guard didDocument.id == did else {
            throw OAuthValidationError.invalidDIDDocument
        }
        if let handle {
            let claimedHandles = didDocument.alsoKnownAs ?? []
            guard claimedHandles.contains(where: {
                $0.caseInsensitiveCompare("at://\(handle)") == .orderedSame
            }) else {
                throw OAuthValidationError.invalidDIDDocument
            }
        }

        guard
            let service = didDocument.service?.first(where: {
                $0.id == "#atproto_pds" ||
                    $0.id == "\(did)#atproto_pds"
            }),
            service.type == "AtprotoPersonalDataServer",
            let pdsURL = secureOrigin(service.serviceEndpoint)
        else {
            throw OAuthValidationError.missingPDS
        }

        let resourceURL = pdsURL.appending(
            path: ".well-known/oauth-protected-resource",
            directoryHint: .notDirectory
        )
        let resource: ProtectedResourceMetadata = try await http.getJSON(
            resourceURL
        )
        guard
            resource.authorizationServers.count == 1,
            let issuer = secureOrigin(resource.authorizationServers[0])
        else {
            throw OAuthValidationError.invalidProtectedResourceMetadata
        }

        let metadataURL = issuer.appending(
            path: ".well-known/oauth-authorization-server",
            directoryHint: .notDirectory
        )
        let server: AuthorizationServerMetadata = try await http.getJSON(
            metadataURL
        )
        try validate(server: server, expectedIssuer: issuer)

        return Result(
            inputIdentifier: identifier,
            did: did,
            handle: handle,
            pdsURL: pdsURL,
            server: server
        )
    }

    func validateAccount(_ did: String, pdsURL: URL, issuer: URL) async throws {
        let document = try await resolveDID(did)
        guard
            document.id == did,
            let service = document.service?.first(where: {
                $0.id == "#atproto_pds" || $0.id == "\(did)#atproto_pds"
            }),
            secureOrigin(service.serviceEndpoint) == pdsURL
        else {
            throw OAuthValidationError.accountMismatch
        }

        let resource: ProtectedResourceMetadata = try await http.getJSON(
            pdsURL.appending(
                path: ".well-known/oauth-protected-resource",
                directoryHint: .notDirectory
            )
        )
        guard resource.authorizationServers.count == 1,
              secureOrigin(resource.authorizationServers[0]) == issuer else {
            throw OAuthValidationError.accountMismatch
        }
    }

    private func resolveHandle(_ handle: String) async throws -> String {
        if let url = URL(string: "https://\(handle)/.well-known/atproto-did"),
           let response = try? await http.get(url),
           response.statusCode == 200,
           let did = String(data: response.data, encoding: .utf8)?
               .trimmingCharacters(in: .whitespacesAndNewlines),
           did.hasPrefix("did:") {
            return did
        }

        var components = URLComponents(
            string: "https://cloudflare-dns.com/dns-query"
        )!
        components.queryItems = [
            URLQueryItem(name: "name", value: "_atproto.\(handle)"),
            URLQueryItem(name: "type", value: "TXT"),
        ]
        let response = try await http.get(
            components.url!,
            headers: ["Accept": "application/dns-json"]
        )
        guard response.statusCode == 200,
              let object = try JSONSerialization.jsonObject(
                with: response.data
              ) as? [String: Any],
              let answers = object["Answer"] as? [[String: Any]] else {
            throw OAuthValidationError.handleCouldNotBeResolved
        }
        for answer in answers {
            guard var value = answer["data"] as? String else { continue }
            value = value.trimmingCharacters(in: CharacterSet(charactersIn: "\""))
            if value.hasPrefix("did=") {
                let did = String(value.dropFirst(4))
                if did.hasPrefix("did:") {
                    return did
                }
            }
        }
        throw OAuthValidationError.handleCouldNotBeResolved
    }

    private func resolveDID(_ did: String) async throws -> DIDDocument {
        let url: URL
        if did.hasPrefix("did:plc:") {
            guard let value = URL(string: "https://plc.directory/\(did)") else {
                throw OAuthValidationError.invalidDIDDocument
            }
            url = value
        } else if did.hasPrefix("did:web:") {
            let methodSpecificID = String(did.dropFirst("did:web:".count))
            let segments = methodSpecificID.split(separator: ":").map(String.init)
            guard let encodedHost = segments.first,
                  let host = encodedHost.removingPercentEncoding,
                  isValidWebDIDHost(host) else {
                throw OAuthValidationError.invalidDIDDocument
            }
            let path = segments.dropFirst().map {
                $0.removingPercentEncoding ?? $0
            }
            var components = URLComponents()
            components.scheme = "https"
            components.host = host
            components.path = path.isEmpty
                ? "/.well-known/did.json"
                : "/" + path.joined(separator: "/") + "/did.json"
            guard let value = components.url else {
                throw OAuthValidationError.invalidDIDDocument
            }
            url = value
        } else {
            throw OAuthValidationError.invalidDIDDocument
        }
        return try await http.getJSON(url)
    }

    private func validate(
        server: AuthorizationServerMetadata,
        expectedIssuer: URL
    ) throws {
        guard
            secureOrigin(server.issuer) == expectedIssuer,
            server.issuer == expectedIssuer,
            server.responseTypesSupported.contains("code"),
            server.grantTypesSupported.contains("authorization_code"),
            server.grantTypesSupported.contains("refresh_token"),
            server.codeChallengeMethodsSupported.contains("S256"),
            server.tokenEndpointAuthMethodsSupported.contains("none"),
            server.scopesSupported.contains("atproto"),
            server.authorizationResponseIssuerParameterSupported,
            server.requirePushedAuthorizationRequests,
            server.dpopSigningAlgorithmsSupported.contains("ES256"),
            server.clientIDMetadataDocumentSupported,
            isSecureEndpoint(server.authorizationEndpoint),
            isSecureEndpoint(server.tokenEndpoint),
            isSecureEndpoint(server.pushedAuthorizationRequestEndpoint)
        else {
            throw OAuthValidationError.invalidAuthorizationServerMetadata
        }
    }

    private func isValidHandle(_ value: String) -> Bool {
        value.count <= 253 &&
            value.range(
                of: #"(?i)^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z](?:[a-z0-9-]{0,61}[a-z0-9])?$"#,
                options: .regularExpression
            ) != nil
    }

    private func isValidWebDIDHost(_ value: String) -> Bool {
        !value.isEmpty &&
            !value.contains("/") &&
            !value.contains("@") &&
            !value.contains("\\")
    }

    private func isSecureEndpoint(_ url: URL) -> Bool {
        url.scheme == "https" && url.host != nil && url.user == nil && url.password == nil
    }

    private func secureOrigin(_ value: String) -> URL? {
        guard let url = URL(string: value) else { return nil }
        return secureOrigin(url)
    }

    private func secureOrigin(_ url: URL) -> URL? {
        guard isSecureEndpoint(url) else { return nil }
        var components = URLComponents()
        components.scheme = "https"
        components.host = url.host
        components.port = url.port
        return components.url
    }
}
